tensorflow-framework/src/main/java/org/tensorflow/framework/metrics/Metric.java

```diff
@@ -0,0 +1,284 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+=======================================================================*/
+package org.tensorflow.framework.metrics;
+
+import org.tensorflow.ExecutionEnvironment;
+import org.tensorflow.Operand;
+import org.tensorflow.framework.initializers.Initializer;
+import org.tensorflow.framework.metrics.impl.MetricVariable;
+import org.tensorflow.op.Op;
+import org.tensorflow.op.Ops;
+import org.tensorflow.op.core.Variable;
+import org.tensorflow.types.family.TNumber;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * Base class for Metrics
+ *
+ * @param <U> The data type for the metric values
+ * @param <T> The data type for the metric result
+ */
+public abstract class Metric<U extends TNumber, T extends TNumber> {
+
+  /** variables are stored by ExecutionEnvironment, and then by an identifier name */
+  protected static Map<ExecutionEnvironment, Map<String, MetricVariable<? extends TNumber>>>
+      variableMap = new WeakHashMap<>();
+  /** The TensorFlow Ops */
+  private final Ops tf;
+  /** The random number generator seed value */
+  private final long seed;
+
+  /** The name for this metric. Defaults to {@link Class#getSimpleName()}. */
+  private final String name;
+
+  /**
+   * Creates a Metric with a name of {@link Class#getSimpleName()}
+   *
+   * @param tf the TensorFlow Ops
+   * @param seed the seed for random number generation. An initializer created with a given seed
+   *     will always produce the same random tensor for a given shape and data type.
+   */
+  protected Metric(Ops tf, long seed) {
+    this(tf, null, seed);
+  }
+
+  /**
+   * Creates a Metric
+   *
+   * @param tf the TensorFlow Ops
+   * @param name the name for this metric. If null, name defaults to {@link Class#getSimpleName()}.
+   * @param seed the seed for random number generation. An initializer created with a given seed
+   *     will always produce the same random tensor for a given shape and data type.
+   */
+  protected Metric(Ops tf, String name, long seed) {
+    if (!tf.scope().env().isGraph())
+      throw new IllegalArgumentException("Metrics are required to execute in Graph mode.");
```

Comment:
I normally include everything in braces and have been doing that since my early `C` days. However, someone did a nit on that in one of the previous PRs. I am happy to go back to my original way of doing single line blocks with braces. I find it is easier to read, and creates a lot less work if you have to modify the code later and make it multi-line.

Reply:
Changed.